README.md
# Phigros_Charts_Stats （中文）

![](https://mirrors.creativecommons.org/presskit/buttons/80x15/png/by-nc-nd.png)

**注意：本数据集遵循[署名-非商业性使用-禁止演绎 3.0 未本地化版本 (CC BY-NC-ND 3.0)](https://creativecommons.org/licenses/by-nc-nd/3.0/deed.zh)协议。若转载、复制、发行本作品等，请附上原文出处链接及本声明。**

本仓库总结了Phigros当中每个谱面的一些信息，包括BPM和物量。同时在这里也提供了一些方便扒谱的工具。

想查看任何已经存在的谱面之变动，可以移步到`changes.md`文件中。

如果有任何问题欢迎在答疑区（Issues）提出。

祝扒谱快乐！

## `Transform.py` 使用方法

首先将自己的谱面(.json)重新命名，序列化为从1到n（n为一个文件夹内，不包含子目录的谱面的数量）。

然后启动该软件，直接输入路径和谱面的数量，程序就会寻找谱面物量数和第一时间的BPM值，将其写在文件名称上。方便日后参照[萌娘百科](https://zh.moegirl.org.cn/Phigros/%E8%B0%B1%E9%9D%A2%E4%BF%A1%E6%81%AF)或本仓库的`result.csv`等文献以比对。

若想添加难度，建议直接在文件名称前面添加`EZ-` `HD-` `IN-` `AT-` `LE-` `SP-`。或者在`.json`前面添加`-EZ`等。

## `Chart_Stats.py` 使用方法 

**现在该代码支持自定义谱面文件目录。**

谱面文件的格式为：一个文件夹，里面包含若干个.json格式的谱面文件。（如果要处理自制谱的话还请阁下自己修改代码了）

譬如一个谱面文件是这样的格式：

```
Mobilys\
    EZ-311-144.0.json
    HD-530-144.0.json
    IN-938-144.0.json
    Mobilys.png
    Mobilys.wav
```

首先确保自己的谱面文件是存放在一个文件夹内的。并且格式为`EZ-XXX-YYY.json`。其中`EZ`可以替换成`HD` `IN`等上文提到过的难度，`XXX`为物量，`YYY`为BPM，且不一定为3位数（这里程序会自动处理好）。

那么所有.json的文件就是谱面文件。而该程序所做的就是将全部谱面文件的信息提取并生成一个.csv文件。

该程序现在支持.zip文件，且可以统计四种音符的数量。

## `result.csv` 使用方法 

这是本人已经统计好的物量信息和BPM，仅供参考。由于部分谱面的物量数和BPM相同，暂时无法直接自动处理文件。

值得一提的是，由于有些歌对大多数用户难以输入，本表对其进行了改动。包括但不限于：

- 某些简繁混搭的曲目将全部使用简体中文来表示（《華灯爱》——《华灯爱》）。
- 过于复杂的歌名将只取前面几个单词（《Infinite Enerzy -Overdoze-》——《Infinite Enerzy》）。
- 如果歌名是日文：
  - 带汉字的，将只取汉字部分（《雪降り、メリクリ》——《雪降》）。
  - 纯日文平假名的，使用罗马字（《もぺもぺ》——《Mopemope》）。
  - 纯日文片假名的，使用其英文形式（《ジングルベル　Jingle Bell》——《Jingle Bell》）。
- 如果歌名为希腊文的：
  - 采取转写（《Ποσειδών》——《Poseidon》）。
- 歌名是一片符号的：
  - 有可识别英文单词的，取英文单词（《Cipher》）。
  - 本身就是魔改的，取英文原意（《ρars/ey》——《parsley》）。

- - -

# Phigros_Charts_Stats （English）

![](https://mirrors.creativecommons.org/presskit/buttons/80x15/png/by-nc-nd.png)

**CAUTION: THIS DATASET IS APPLIED BY [Attribution-NonCommercial-NoDerivs 3.0 Unported (CC BY-NC-ND 3.0)](https://creativecommons.org/licenses/by-nc-nd/3.0/). If you copy, reproduce, distribute this work, etc., please attach a link to the original source and this statement.**

To summarize the information of every beatmaps in Phigros, including BPM and note amount. Also, some tools are provided here, which can help you to extract beatmap files more easily.

If you want to have a look at any changes of existing beatmaps, please check `changes.md`.

If you have any problems, feel free to communicate in "Issues".

Have fun digging!

## Usage of `Transform.py`

First, rename your own beatmap files (.json) and serialize them from 1 to n (n is the number of beatmap files in a folder, excluding subdirectories).

Then start the program, directly input the path and the number of beatmap files, the program will find the note amount and the **initial** BPM of each beatmap file, and rename them to `Notes-BPM.json`. It is convenient to refer to [Details of beatmaps](https://zh.moegirl.org.cn/Phigros/%E8%B0%B1%E9%9D%A2%E4%BF%A1%E6%81%AF) or `result.csv` in this repository for comparison in the future.

If you want to add difficulty, it is recommended to add `EZ-` `HD-` `IN-` `AT-` `LE-` `SP-` directly before the file name. Or add `-EZ` in front of `.json`, etc.

## Usage of `Chart_Stats.py`

**Right now this program has supported customized path of beatmaps.**

The format of the beatmap file is: a folder containing several beatmap files in .json format. (If you want to deal with fanmade beatmaps, please modify the source code yourself)

For example, a beatmap file has the following format:

```
Mobilys\
    EZ-311-144.0.json
    HD-530-144.0.json
    IN-938-144.0.json
    Mobilys.png
    Mobilys.wav
```

Firstly make sure your beatmap files are stored in a folder. And the format of every beatmaps is `EZ-XXX-YYY.json`, where `EZ` can be replaced with `HD` `IN` and other difficulties mentioned above, `XXX` is the quantity, `YYY` is the BPM, and it is not necessarily a 3-digit number (the program will handle it automatically here).

Then all .json files are beatmap files. What the program does is to extract the information from beatmap files and generate a summarized .csv file.
This program also supports .zip files right now. And it could summarize the detialed amounts of Tap, Drag, Hold and Flick.

## Usage of `result.csv`

This is the note amount and BPM that I have already summarized, for reference only. Since there are some beatmaps share the same note amount and BPM information, it is temporarily impossible to process the files automatically.

Because the name of beatmaps can be in English, Chinese, Japanese and Greek, even in symbols. I would like to rename them myself, for people easier to look up the imformation of a specific beatmap.

- Some mashups of simplified and traditional Chinese will be replaced to Simplified Chinese completely. e.g. 華灯爱 -> 华灯爱
- Overcomplicated song titles will only be taken the first few words. e.g. Infinite Enerzy -Overdoze- -> Infinite Enerzy
- If the title is mainly in Japanese:
  - If it includes Kanji, only Kanji will be considered. e.g. 雪降り、メリクリ -> 雪降
  - Pure Japanese Hiragana, using Romaji. e.g. もぺもぺ -> Mopemope
  - Pure Katakana, using its English form. e.g. ジングルベル　Jingle Bell -> Jingle Bell
- If the title is mainly in Greek：
  - JUST TRANSLITERATE IT. e.g. Ποσειδών -> Poseidon
- If the title includes some symbols：
  - If there are some meaningful words, only take them. e.g. Cipher
  - If the title is originated from a specific word, the original word will be taken. e.g. ρars/ey -> parsley


changes.md
# 谱面变动相关 / Changes of Beatmaps

本文件只记录已经存在的谱面之变动，不会记录新增谱面。

This file only contain changes of existed beatmaps; any new-added beatmaps won't be mentioned.

## 20220430 - 2.2.0

- Luminescent

Transform.py
#基于python3.10
#代码来自于https://www.bilibili.com/read/cv15503487#reply108566197056
#代码原作者「我觉得你长得很省电」

#加载库
import json
import os

number = 0    #可能把for语句记叉了所以只会用while做循环，定义一个计数变量
workpath = input("请输入铺面位置（绝对路径）：")+"//"
allcount = int(input("请输入总铺面数（即总文件数）："))-1

while number <= int(allcount):
    os.chdir(str(workpath))
    number = number + 1

    try:
        with open("{0}.json".format(number),mode="r") as f:  #打开文件
            s = f.read()
            d = json.loads(s)
            i = d.get("numOfNotes")     #调取物量数据
            j = d.get('judgeLineList')[0].get('bpm')#调取BPM信息
            f.close

        src = os.path.join(workpath,str(number)+'.json')     #记录路径和更改后名字
        #dst = os.path.join(workpath,str(i)+'-'+str(j)+'('+str(number)+')'+'.json')
        dst = os.path.join(workpath, str(i) + '-' + str(j) + '.json')

        os.rename(src,dst)  #更名

        #print(str(number)+".json已处理，结果为："+str(i)+'-'+str(j)+'('+str(number)+')'+'.json')     #输出结果
        print(str(number) + ".json已处理，结果为：" + str(i) + '-' + str(j) + '.json')  # 输出结果

    except:

        print("Error:"+str(number)) 

Chart_Stats.py
# coding=UTF-8

'''
谱面文件的格式为：一个文件夹，里面包含若干个.json格式的谱面文件。
譬如一个谱面文件是这样的格式：

Mobilys\
    EZ-311-144.0.json
    HD-530-144.0.json
    IN-938-144.0.json
    Mobilys.png
    Mobilys.wav

那么所有.json的文件就是谱面文件。而该程序所做的就是将谱面文件的信息提取并生成一个.csv文件。
execute_files()以AT EZ HD IN LE SP为顺序进行整理。
execute_diffs()以EZ HD IN AT LE SP为顺序进行整理。
'''

import csv
from ctypes import sizeof
import os
import time
import json
import jsonpath
import zipfile

headers = ['Name', 'Difficulty', 'Notes', 'BPM', 'Tap', 'Drag', 'Hold', 'Flick']    # 谱面的基本信息：名字、难度、物量、节拍
diffs = ['EZ', 'HD', 'IN', 'AT', 'LE', 'SP']        # 六大难度

totalStats = []                                     # 待写入到文件的总数据

def execute_folders(target):
    #currentPath = os.getcwd()                                       # 获得脚本所在目录
    currentPath = target
    for _, dirs, _ in os.walk(currentPath):                         # 遍历当前目录
        for dir in dirs:                                            # 遍历每一个子目录，其实就是访问一个个歌曲文件夹
            currentSubPath = os.path.join(currentPath, dir)                     # 设定子目录
            for _, _, files in os.walk(currentSubPath):             # 提取该歌曲文件夹内的全部文件
                for diff in diffs:
                    chartData = []
                    isChart = False
                    chartData.append(dir)
                    chartData.append(diff)
                    for file in files:
                        if 'fake' in file:
                            continue
                        if diff in file and ".json" in file:
                            isChart = True
                            dataString = file[3:-5]
                            dataString = dataString.split('-')
                            chartData.append(dataString[0])
                            chartData.append(dataString[1])

                            '''
                            以上操作是通过读取文件名的方式获取谱面的物量和初始BPM信息，接下来的部分是获取每个谱面的详细物量
                            也就是Tap, Drag, Hold, Flick这四种音符各自有多少个
                            '''

                            with open(os.path.join(currentSubPath, file), mode="r") as f:
                                #print("Opening: {0}".format(os.path.join(currentSubPath, file)))
                                fileRead = f.read()
                                jsonLoad = json.loads(fileRead)

                                # 获取谱面物量信息，会形成一个含有 1 2 3 4 这四个数字的列表。这四个数字分别代表上述的四种音符。
                                types_above = jsonpath.jsonpath(jsonLoad, "$.judgeLineList[*].notesAbove[*].type")
                                types_below = jsonpath.jsonpath(jsonLoad, "$.judgeLineList[*].notesBelow[*].type")
                                # print(types_above, types_below)
                                # 有的谱面没有从判定线地下浮上来的音符，因此结果是False．这个时候再直接与数组相加的话会报错
                                types = []
                                if types_above != False:
                                    types += types_above
                                if types_below != False:
                                    types += types_below
                                # 该谱面音符信息综述。统计每个音符在列表当中的次数
                                noteResult = []
                                for i in range(1, 5):
                                    noteResult.append(str(types.count(i)))
                                # 如果统计过程中的音符数与从文件当中获取的音符数不相符，就不会再录入
                                try:
                                    if len(types) != int(dataString[0]):
                                        raise Exception('NoteAmountNotEqual')
                                    chartData = chartData + noteResult
                                except Exception:
                                    print("ERROR: Note amount not Equal (notes != tap+drag+hold+flik)")
                                    print("错误：四种音符数量的总和不等于谱面总物量！")
                                f.close()
                    if isChart:
                        print(chartData)
                        totalStats.append(chartData)
                        
    #print(totalStats)

    with open('result.csv', 'w', newline='', encoding='utf-8') as f:
        f_csv = csv.writer(f)
        f_csv.writerow(headers)
        f_csv.writerows(totalStats)

def execute_zips(target):
   #currentPath = os.getcwd()
    currentPath = target
    for _, _, files in os.walk(currentPath):
        for file in files:  #打开当前文件夹的zip文件
            filePath = currentPath + os.sep + file
            if zipfile.is_zipfile(filePath):    #检验是不是zip文件
                with zipfile.ZipFile(filePath, 'r') as z:   #打开zip文件
                    nameList = z.namelist()     #压缩文件内的文件名目录
                    for diff in diffs:          #按照难度进行排序
                        beatmap_gotten = False  #是否得到当前难度的谱面？
                        for fileInside in nameList: #对每个压缩包内文件进行遍历
                            fileInsideName = fileInside.replace(nameList[0], '')    #将前面的路径删掉
                            if '.json' not in fileInsideName:   #.json不在文件名内，说明不是官方谱面文件
                                continue
                            if diff not in fileInsideName:  #难度不在文件名内，说明不是谱面文件
                                continue
                            if 'fake' in fileInsideName:    #fake特殊标记在文件名内，自动忽略
                                continue
                            beatmap_gotten = True
                            chartData = []
                            chartData.append(file.replace(".zip", ''))  #曲名取自zip文件的压缩包名
                            chartData.append(diff)
                            dataString = fileInsideName[3:-5]
                            dataString = dataString.split('-')
                            chartData.append(dataString[0])
                            chartData.append(dataString[1])

                            with z.open(fileInside, mode="r") as f:
                                #print("Opening: {0}".format(fileInside))
                                fileRead = f.read().decode('utf-8')
                                jsonLoad = json.loads(fileRead)

                                #print(jsonLoad)
                                # 获取谱面物量信息，会形成一个含有 1 2 3 4 这四个数字的列表。这四个数字分别代表上述的四种音符。
                                types_above = jsonpath.jsonpath(jsonLoad, "$.judgeLineList[*].notesAbove[*].type")
                                types_below = jsonpath.jsonpath(jsonLoad, "$.judgeLineList[*].notesBelow[*].type")
                                # print(types_above, types_below)
                                # 有的谱面没有从判定线地下浮上来的音符，因此结果是False．这个时候再直接与数组相加的话会报错
                                types = []
                                if types_above != False:
                                    types += types_above
                                if types_below != False:
                                    types += types_below
                                # 该谱面音符信息综述。统计每个音符在列表当中的次数
                                noteResult = []
                                for i in range(1, 5):
                                    noteResult.append(str(types.count(i)))
                                # 如果统计过程中的音符数与从文件当中获取的音符数不相符，就不会再录入
                                try:
                                    if len(types) != int(dataString[0]):
                                        raise Exception('NoteAmountNotEqual')
                                    chartData = chartData + noteResult
                                except Exception:
                                    print("ERROR: Note amount not Equal (notes != tap+drag+hold+flik)")
                                    print("错误：四种音符数量的总和不等于谱面总物量！")
                                f.close()

                        if beatmap_gotten:
                            print(chartData)
                            totalStats.append(chartData)
                    z.close()

    #print(totalStats)
    with open('result.csv', 'w', newline='', encoding='utf-8') as f:
        f_csv = csv.writer(f)
        f_csv.writerow(headers)
        f_csv.writerows(totalStats)
        f.close()

if __name__ == "__main__":
    #debug__=os.getcwd()
    _method = '0'
    targetPath = input("Beatmap Path 谱面目录:")

    while _method not in ['f', 'F', 'z', 'Z']:
        _method = input('''
Please type the method of extracting beatmap files.
"Z" means to load every .zip files, and "F" means to load every folders.
This program is not case-sensitive.
请输入提取谱面文件的方式字：Z代表提取压缩文件，F代表提取文件夹。该操作大小写不敏感。
''')

    initTime=time.time()
    if _method in ['f', 'F']:
        execute_folders(targetPath)
    elif _method in ['Z', 'z']:
        execute_zips(targetPath)
    print("Found {0} beatmaps.".format(len(totalStats)))
    print("Running time: {0}".format(time.time()-initTime))
    